azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+            assert cb_mock.call_count == 1
+            assert cb_mock.call_args == mocker.call(op=worker_op, error=error)
+
+            # The original op that spawned the worker is also completed
+            assert op.complete.call_count == 1
+            assert op.complete.call_args == mocker.call(error=error)
+
+    @pytest.mark.describe("{} - .complete()".format(op_class_under_test.__name__))
+    class OperationCompleteTests(OperationTestConfigClass):
+        @pytest.fixture(params=["Successful completion", "Completion with error"])
+        def error(self, request, arbitrary_exception):
+            if request.param == "Successful completion":
+                return None
+            else:
+                return arbitrary_exception
+
+        @pytest.mark.it(
+            "Marks the operation as completed by setting the 'completed' attribute to True"
+        )
+        def test_marks_complete(self, op, error):
+            assert not op.completed
+            op.complete(error=error)
+            assert op.completed
+
+        @pytest.mark.it(
+            "Sends an OperationError to the background exception handler if the operation has already been completed"
+        )
+        def test_already_complete(self, mocker, op, error):
+            mocker.spy(handle_exceptions, "handle_background_exception")
+            op.complete(error=error)
+            assert op.completed
+            assert handle_exceptions.handle_background_exception.call_count == 0
+            op.complete(error=error)
+            assert handle_exceptions.handle_background_exception.call_count == 1
+            assert (
+                type(handle_exceptions.handle_background_exception.call_args[0][0])
+                is pipeline_exceptions.OperationError
+            )
+
+        @pytest.mark.it(
+            "Triggers and removes callbacks that have been added to the operation according to LIFO order"
+        )
+        def test_trigger_callbacks(self, mocker, cls_type, init_kwargs, error):
+            mocker.spy(handle_exceptions, "handle_background_exception")
+
+            # Set up callback mocks
+            cb1_mock = mocker.MagicMock()
+            cb2_mock = mocker.MagicMock()
+            cb3_mock = mocker.MagicMock()
+
+            def cb1(op, error):
+                # Callback 2 and 3 have already been triggered
+                assert cb1_mock.call_count == 1
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 0
+
+            def cb2(op, error):
+                # Callback 3 has already been triggered, but Callback 1 has not
+                assert cb1_mock.call_count == 0
+                assert cb2_mock.call_count == 1
+                assert cb3_mock.call_count == 1
+                assert len(op.callbacks) == 1
+
+            def cb3(op, error):
+                # Callback 3 has been triggered, but no others have been.
```

Comment:
>allback 3 has been triggered, but no others have been. [](start = 19, length = 54)

Nit: Comments on these 3 callbacks are not symmetrical.  First 2 comments describe pre-conditions (state before the callback is called) and last commend describes post-condition (state after the callback is called).